### PR TITLE
Update integration test setup

### PR DESCRIPTION
**Pull request by bfbachmann**
## Problem
We want to replace the k3d downstream cluster creation using k3d with v2prov cluster creation. This is because the existing k3d setup causes conflicts between concurrent Drone builds.
 
## Solution
- Update downstream cluster creation in tests/v2/integration/setup so it uses v2prov instead of k3d.
- Update scripts/test so it matches what is being done in scripts/provisioning-tests
- Clean up error handling and logging in tests/v2/integration/setup
- Remove dead code in tests/framework/clients/k3d
 
## Testing
Run all unit and integration tests using `make test`.

### Manual Testing
N/A

### Automated Testing
Already covered by CI.

### CODE REVIEW

File: scripts/test
Comment (bfbachmann) [inline scripts/test:1]: I've updated this file to more closely match what we're doing in `scripts/provisioning-tests`. Pretty much all the changes in here were copied from that script.

File: tests/framework/clients/dynamic/dynamic.go
Comment (bfbachmann) [inline tests/framework/clients/dynamic/dynamic.go:26]: This log is very spammy during integration tests. I figure it should be a debug log.

File: tests/framework/clients/k3d/config.go
Comment (bfbachmann) [inline tests/framework/clients/k3d/config.go:1]: I removed this file and the one below because none of the exported functions or variables inside them are used anymore after the transition away from k3d.

File: tests/framework/extensions/users/users.go
Comment (bfbachmann) [inline tests/framework/extensions/users/users.go:305]: This test was flaking out often in CI. I figured we can afford to wait an extra 500ms to give it a better chance at passing.

File: tests/framework/pkg/config/config.go
Comment (bfbachmann) [inline tests/framework/pkg/config/config.go:98]: I'd prefer to return an error over panicking here.

File: tests/v2/integration/setup/main.go
Comment (bfbachmann) [inline tests/v2/integration/setup/main.go:1]: This is where the meat of the change is. Here, I'm replacing the k3d cluster creation with v2prov.

File: tests/v2prov/defaults/defaults.go
Comment (bfbachmann) [inline tests/v2prov/defaults/defaults.go:6]: I've bumped this test image version because the newest version has both amd64 and arm64 images, and we need both for CI.

File: Dockerfile.dapper
Comment (bfbachmann) [inline Dockerfile.dapper:113]: We don't need to use the host network inside the Dapper container anymore. This was originally done because it was necessary for using k3d, but now that we've replaced it we don't need this anymore and it only causes problems.

File: tests/v2prov/defaults/defaults.go
Comment (bfbachmann) [reply tests/v2prov/defaults/defaults.go:6]: This new test image also makes it so our tests finally work on Rancher Desktop for Mac!

File: scripts/test
Comment (jakefhyde) [inline scripts/test:129]: Can we give this a more descriptive name? It doesn't have to be verbose (`i` would be fine, or `attempt`, etc.), it's just not obvious without looking at the associated if statements. It might also be a good idea to make the logic generic, like so:
```
maxWaitSeconds=300 # 5 minutes
sleepSeconds=2
i=0
echo "Waiting up to 5 minutes for rancher-webhook deployment"
while ! kubectl --kubeconfig /etc/rancher/k3s/k3s.yaml rollout status -w -n cattle-system deploy/rancher-webhook &>/dev/null; do
  if [[ "$i" == $((maxWaitSeconds/sleepSeconds)) ]]; then
    echo "rancher-webhook was not available within $((maxWaitSeconds/60)) minutes. failing fast"
    exit 1
  fi
  if [[ $((i/(60/sleepSeconds))) == 0 ]] && [[ "$i" != "0" ]]; then
    echo "rancher-webhook was not ready within $((i/(60/sleepSeconds))) minutes"
  fi
    sleep $sleepSeconds
    i=$((i + 1))
done
```
Just a thought, mostly because changing in the future would require some back of the napkin math I wouldn't want to do lol

**Review by jakefhyde — CHANGES_REQUESTED** [review]
Few nits, nice job overall

File: tests/framework/clients/k3d/k3d.go
Comment (igomez06) [inline tests/framework/clients/k3d/k3d.go:28]: I don't think we should delete k3d functionality just because we won't use it in the integration setup. Let's keep this for maybe a future usage of some sort.

File: tests/framework/pkg/config/config.go
Comment (igomez06) [reply tests/framework/pkg/config/config.go:98]: Do you have a specific reason for this? Ideally we kept the config file stuff as panics simply because if these steps failed then we should panic since the framework needs to config files to be able to work. I'm open to changing this, just wanted an explanation.

File: tests/v2/integration/setup/main.go
Comment (igomez06) [reply tests/v2/integration/setup/main.go:29]: I agree

File: tests/v2/integration/setup/main.go
Comment (igomez06) [inline tests/v2/integration/setup/main.go:84]: One of the things that was an advantage for using the k3d cluster is being able to use the latest version of rancher-agent. What version of rancher-agent does the cluster now run? Does it not matter?

**Review by igomez06 — CHANGES_REQUESTED** [review]
Overall good, would like an explanation on why the agent tag stuff will be removed.

File: tests/v2/integration/setup/main.go
Comment (igomez06) [inline tests/v2/integration/setup/main.go:55]: also one nit pick it's go standards to have errors begin in lower case.

File: tests/v2/integration/setup/main.go
Comment (bfbachmann) [reply tests/v2/integration/setup/main.go:84]: As far as I know, the system agent version used in tests will be taken from `package/Dockerfile` (see [here](https://github.com/rancher/rancher/pull/41771/files#diff-d7ad7c7666dae3e86d4d3a850f7c4df6dcde86454e9995001ea017e44a3d35d5R46)). This should just be the latest system agent version. The old code was just using the value from the envvar AGENT_TAG, which I think has the same result.

Comment (igomez06) [reply tests/v2/integration/setup/main.go:84]: It's not the same result, that's why here: https://github.com/rancher/rancher/blob/cbf32214807bdb2ba6795b61f6bb79c6138beb66/scripts/ci#L14 we're building the docker images before validating. This build is happening from the latest commit, hence why in the setup it was updating the `management.cattle.io.setting` to use the latest image. That way changes to the agent image can be tested within the same PR.

Comment (igomez06) [reply tests/v2/integration/setup/main.go:84]: @cmurphy weren't you mentioning a possible use case to write integration tests against the "latest" agent changes?

File: tests/v2/integration/setup/main.go
Comment (snasovich) [reply tests/v2/integration/setup/main.go]: @igomez06 , this standard is something that's often misinterpreted - it only applies to actual error messages and not to text that gets logged, https://github.com/go-lang-plugin-org/go-lang-idea-plugin/issues/2151 has a good summary on this.

Comment (igomez06) [reply tests/v2/integration/setup/main.go]: @snasovich sounds good, thanks for the summary! :grin:

**Review by jakefhyde — DISMISSED** [review]

**Comment by jakefhyde** [comment]
Looks like CI is failing on `TestRTBTestSuite/TestCRTBRoleTemplateInheritance`. If you run CI and it passes, we should create a github issue to track it.

File: tests/framework/clients/k3d/config.go
Comment (igomez06) [reply tests/framework/clients/k3d/config.go:1]: I understand that, but imo that doesn't mean we should completely remove the k3d support. This wasn't just meant to be used in the integration drone pipeline. This code existed before we even had the go integration tests in the pipeline. I definitely disagree with having to remove all this code.

File: tests/v2/integration/setup/main.go
Comment (bfbachmann) [reply tests/v2/integration/setup/main.go:84]: Thanks for the feedback. I've updated the test setup so it now creates an additional in-cluster registry and copies test images (i.e. `rancher-agent:master-head`) to it before using those images in the downstream clusters. Moving forward (but in a separate PR), I'll also update the `provisioning-test` setup to do the same thing so we're testing against the current rancher agent in those tests as well.

Comment (igomez06) [reply tests/v2/integration/setup/main.go:84]: Sounds good, but from what I remember the locally created rancher-agent image has the commit as the tag, rather it just simply being `master-head` hence why I was doing this
```
var (
 	agentTag         = os.Getenv("AGENT_TAG")
 	masterAgentImage = "rancher/rancher-agent:" + agentTag
 )
```

**Review by igomez06 — APPROVED** [review]
Looks good to me. @snasovich Since there a lot of changes done to the scripts, should we get a review from Seb ?

**Review by igomez06 — APPROVED** [review]
Looks go to me, but should probably get @superseb to give a chance to review the updates on the drone scripts

File: scripts/validate
Comment (superseb) [inline scripts/validate:7]: Does this need to go? Just remove it and add why

Comment (bfbachmann) [reply scripts/validate:7]: Oh good catch, thank you. I didn't mean to remove it, I just commented it out so it wouldn't run while I was testing locally. I'll undo this change.

Comment (bfbachmann) [reply scripts/validate]: @superseb Fixed. Ty

**Review by igomez06 — APPROVED** [review]
LGTM

File: pkg/portforward/portforward.go
Comment (cmurphy) [inline pkg/portforward/portforward.go:1]: It looks like this is only being used in the tests, so it should be moved under `tests/` instead of being included in the main rancher package.

**Review by cmurphy — CHANGES_REQUESTED** [review]
With this change, what's the difference between the regular integration tests and the provisioning tests? Should they be consolidated?

One minor issue noted inline, otherwise lgtm

**Review by superseb — APPROVED** [review]

File: pkg/portforward/portforward.go
Comment (bfbachmann) [reply pkg/portforward/portforward.go:1]: Thanks, I moved the portforward package into `/tests`.

As far as I know, the integration and provisioning test suites are similar in that they create downstream clusters the same way, so it might be worth consolidating them in the future. Unlike the provisioning test suit, however, the integration tests seem to be testing a lot of functionality in the Steve API that is not specific to cluster provisioning. I think it would be a lot of work to consolidate them, but it's definitely something we should think about moving forward.

**Review by cmurphy — APPROVED** [review]
lgtm, can you make sure to squash these commits before merging?

**Pull request merged**
